### ISSUE TITLE
Test if target is a valid host

### ISSUE THREAD

**Issue by J0WI** [EFForg/https-everywhere#10877]
See #10869

**Comment by cschanaj** [comment]
@J0WI thank you for opening this issue. I would like to add some details here.

In #10869 I checked if a `target `is valid by matching it against the [`Public Suffix List`](https://publicsuffix.org/). While this test does not guarantee each target has a working DNS record, it has the advantage that no network request is required to perform the checking. 

It was found that [`AM-Best-Company.xml#L11`](https://github.com/EFForg/https-everywhere/blob/b7650c06f79e61eefdc34f3b7e6a686d990b8e2b/src/chrome/content/rules/AM-Best-Company.xml#L11) with a invalid `target `passed the comprehensive fetch test and got re-activated in #10538. Despite @Bisaloo's assertion, I would like to hear from @Hainish for an explaination.

**Comment by cschanaj** [comment]
@Hainish @J0WI I have implemented the logic which check each `target` is having a valid `tld` in the `rule-test` in cschanaj/https-everywhere/pull/7. Besides, I noted that the following type of rewrite will pass the `fetch-test` without any Travis complaint

```xml
<ruleset name="cschanaj">
  <target host="dns-record-does-not-exist.example.com" />
  
  <rule from="^http://dns-record-does-not-exist\.example\.com/"
          to="https://example.com/" />
</ruleset>
```

Is this an intended behavior? Should we check each `target` for an existing DNS record? See cschanaj/https-everywhere/pull/6

**Comment by cschanaj** [comment]
@Hainish can you have a look on https://github.com/EFForg/https-everywhere/issues/10877#issuecomment-315598041 and #11187 ?

**Comment by Hainish** [comment]
Not quite sure how [`AM-Best-Company.xml#L11`](https://github.com/EFForg/https-everywhere/blob/b7650c06f79e61eefdc34f3b7e6a686d990b8e2b/src/chrome/content/rules/AM-Best-Company.xml#L11) got re-enabled.  Checking targets against the Public Suffix List seems reasonable.

**Comment by cschanaj** [comment]
@Hainish  For any `target` if HTTPS connection works, the fetch test will ignore any result for the HTTP connection, see [`check_rules.py#L131`](https://github.com/EFForg/https-everywhere/blob/master/test/rules/src/https_everywhere_checker/check_rules.py#L131). In particular, it will ignore the `could not resolve host` error. As a result, any rewrite in the form of https://github.com/EFForg/https-everywhere/issues/10877#issuecomment-315598041 will always pass the fetch test.

It seems to be a bug to me. A quick fix is to throw an error if HTTP cannot be resolved. Can you confirm this?

**Comment by cschanaj** [comment]
All ruleset in the code base will pass the proposed `rule-test` once #11263 is merged, it will be also safe to merge #11187 (hopefully before next release).
